File: typescript/src/sprites/explosion.ts
import P5 from 'p5';
import Particle from './particle';

export default class Explosion {
    particles: Particle[] = [];

    constructor(p5: P5, position: P5.Vector) {
        for (let index = 0; index < 8; index++) {
            this.particles.push(
                new Particle(p5, position.copy(), p5.random(1, 4))
            );
        }
    }

    update(): boolean {
        this.particles = this.particles.filter((particle) => particle.update());

        return this.particles.length > 0;
    }

    draw() {
        for (const particle of this.particles) {
            particle.draw();
        }
    }
}

File: rust/src/sprites/explosion.rs
use web_sys::CanvasRenderingContext2d;

use crate::utils::config::Config;

use super::{
    particle::Particle,
    sprite::{CanvasDimension, Spritable, SpriteData},
};

pub struct Explosion {
    particles: Vec<Particle>,
    pub is_faded: bool,
}

impl Explosion {
    pub fn new(sprite_data: SpriteData, canvas: CanvasDimension) -> Explosion {
        let config = Config::new();

        let mut particles = Vec::new();

        for _count in 0..config.sprites.explosion.count_particules {
            particles.push(Particle::new(sprite_data, canvas));
        }

        Explosion {
            particles,
            is_faded: false,
        }
    }
    pub fn update(&mut self) {
        for index in (0..self.particles.len()).rev() {
            self.particles[index].update();

            if self.particles[index].is_faded {
                self.particles.remove(index);
            }
        }

        if self.particles.is_empty() {
            self.is_faded = true;
        }
    }

    pub fn draw(&self, canvas: CanvasRenderingContext2d) {
        for particle in &self.particles {
            particle.draw(canvas.clone());
        }
    }
}
